aten/src/ATen/native/Linear.cpp

```diff
@@ -336,36 +336,36 @@ Tensor einsum(std::string equation, TensorList operands) {
   for (std::size_t i = 0; i < operands.size(); ++i) {
     std::vector<int64_t> perm_shape(perm_index, -1);
     std::vector<int64_t> label_dim(TOTAL_LABELS, -1);
-    const std::vector<int> labels = op_labels[i];
     Tensor operand = operands[i];
-    const auto sizes = operand.sizes();
-    std::size_t j = 0;
+    const auto labels = op_labels[i];
+    const auto original_sizes = operand.sizes();
 
-    for (int label : labels) {
+    std::size_t j = 0;
+    for (auto label : labels) {
       if (label == ELLIPSIS) {
-        // Add missing dimensions under ellipsis
-        int64_t num_dim_diff =
-            ell_num_dim - (operand.dim() - labels.size() + 1);
```

Comment:
The bug was here. I was using the current operand size to compute how many dimensions were covered by the ellipsis. But if I happened to take a diagonal before the ellipsis the dimensions of operand would change and this value would be wrong.